Refactor testconn to use get_sqla_engine

thoralf-gutierrez opened the pull request:
### CATEGORY

Choose one

- [ ] Bug Fix
- [ ] Enhancement (new features, refinement)
- [x] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Creating the database connection will actually go through the get_sqla_engine path so that's what should be used to test the connection. This was triggered by realizing that testconn was not using the DB_CONNECTION_MUTATOR hook.

See https://github.com/apache/incubator-superset/issues/7668

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TEST PLAN
<!--- What steps should be taken to verify the changes -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

john-bodley commented on superset/views/core.py:
```diff
@@ -1796,46 +1795,29 @@ def testconn(self):
-            database = None
+
+            # extras is sent as json, but required to be a string in the Database model
+            extra_json = request.json.get('extras', {})
```
I would probably just make this one line called `extras` (inline with the JSON name), i.e., 

```
extras = json.dumps(request.json.get('extras', {}))
```

john-bodley commented on superset/views/core.py:
```diff
@@ -1796,46 +1795,29 @@ def testconn(self):
+
             if db_name:
-                database = (
+                existing_database = (
```
Would you be able to clarify the difference between `existing_database` and `database` (defined below)?

john-bodley reviewed (commented):
Thanks for refactoring the code.

thoralf-gutierrez commented:
Thanks for the early comments @john-bodley -- I fixed how I was creating the database (needed to use `set_sqlalchemy_uri`) and now the test passes. I think I've also addressed your comments. Can you take a second look?

john-bodley commented on superset/views/core.py:
```diff
@@ -2018,53 +2017,38 @@ def testconn(self):
-                connect_args["configuration"] = configuration
+            # this is the database instance that will be tested
+            database = models.Database(
+                extra=extra_str, impersonate_user=impersonate_user
```
Given that `extra_str` isn't used elsewhere I would probably just have all the content/context in one place, 

```
database = models.Database(
    extra=json.dumps(request.json.get("extras", {})), 
    impersonate_user=impersonate_user,
)
```

thoralf-gutierrez replied:
:+1: Also went ahead and did the same for other variables.

Commit:
Refactor testconn to use get_sqla_engine

Creating the database connection will actually go through the get_sqla_engine path so that's what should be used to test the connection. This was triggered by realizing that testconn was not using the DB_CONNECTION_MUTATOR hook.

See #7668

john-bodley commented:
@thoralf-gutierrez would you mind removing the `[WIP]` prefix if this PR is no longer work-in-progress?
